### PR TITLE
[nrfconnect] Added support for updating network core using new DFU Target library.

### PR DESCRIPTION
**Pull request by ArekBalysNordic**
#### Problem
Old DFU Target API did not allow to perform a firmware update for app core and net core simultaneously. 
DFU Target API needs to be updated and utilized in the OTA process for nrfconnect boards.

#### Change overview
* Updated NCS revision to cc0412169 to get new DFU Target revision.
* Utilized new DFU Target API.
* Handled both application and network core image swap for NRF5340.

#### Testing
* Tested manually on NRF5340DK
* CI tested

### CODE REVIEW

File: src/platform/nrfconnect/OTAImageProcessorImpl.cpp [hunk 1]
```diff
@@ -45,23 +46,12 @@ CHIP_ERROR OTAImageProcessorImpl::PrepareDownloadImpl()
...
-
-CHIP_ERROR OTAImageProcessorImpl::Abort()
-{
-    return System::MapErrorZephyr(dfu_target_reset());
-}
-
-CHIP_ERROR OTAImageProcessorImpl::Apply()
-{
-    ReturnErrorOnFailure(System::MapErrorZephyr(dfu_target_done(true)));
-
```
Comment (Damian-Nordic) [inline src/platform/nrfconnect/OTAImageProcessorImpl.cpp:64]: Why was this changed? The reboot should happen only after `Apply`, not after finalizing the download.

File: src/platform/nrfconnect/OTAImageProcessorImpl.cpp [hunk 2]
```diff
@@ -129,8 +155,15 @@ CHIP_ERROR OTAImageProcessorImpl::ProcessHeader(ByteSpan & block)
 
     if (mContentHeaderParser.IsInitialized() && !block.empty())
     {
-        OTAImageContentHeader header = {};
-        CHIP_ERROR error             = mContentHeaderParser.AccumulateAndDecode(block, header);
+        CHIP_ERROR error = mContentHeaderParser.AccumulateAndDecode(block, mContentHeader);
+        ChipLogDetail(SoftwareUpdate, "Found following DFU Images:");
```
Comment (Damian-Nordic) [inline src/platform/nrfconnect/OTAImageProcessorImpl.cpp:159]: This should happen after validating the return code.

File: src/platform/nrfconnect/OTAImageProcessorImpl.cpp [hunk 3]
```diff
@@ -129,8 +155,15 @@ CHIP_ERROR OTAImageProcessorImpl::ProcessHeader(ByteSpan & block)
...
+        for (uint8_t i = 0; i < mContentHeader.kMaxFiles; i++)
+        {
+            ChipLogDetail(SoftwareUpdate, "[%d]: \n \
```
Comment (Damian-Nordic) [inline src/platform/nrfconnect/OTAImageProcessorImpl.cpp:162]: Please try not to use multi-line logs, they are more difficult to parse in testing. Also, you can use `%" PRIu32`, instead of "%d" as there are compiler warnings that validate if format string matches the provided data types.

File: src/platform/nrfconnect/OTAImageProcessorImpl.cpp [hunk 4]
```diff
@@ -76,23 +66,59 @@ CHIP_ERROR OTAImageProcessorImpl::Apply()
...
+        {
+            // DFU target library buffers data internally, so do not clone the block data.
+            error = System::MapErrorZephyr(dfu_target_write(block.data(), block.size()));
+        }
+        ChipLogDetail(SoftwareUpdate, "Written %llu/%u Bytes", mParams.downloadedBytes,
+                      mContentHeader.mFiles[mCurrentImage].mFileSize);
     }
 
     // Report the result back to the downloader asynchronously.
     return DeviceLayer::SystemLayer().ScheduleLambda([this, error, block] {
```
Comment (Damian-Nordic) [inline src/platform/nrfconnect/OTAImageProcessorImpl.cpp:119]: Remove `block`.

File: src/platform/nrfconnect/OTAImageProcessorImpl.cpp [hunk 5]
```diff
@@ -45,23 +46,12 @@ CHIP_ERROR OTAImageProcessorImpl::PrepareDownloadImpl()
     ReturnErrorOnFailure(System::MapErrorZephyr(dfu_target_mcuboot_set_buf(mBuffer, sizeof(mBuffer))));
     ReturnErrorOnFailure(System::MapErrorZephyr(dfu_target_reset()));
 
-    return System::MapErrorZephyr(dfu_target_init(DFU_TARGET_IMAGE_TYPE_MCUBOOT, /* size */ 0, nullptr));
+    // Initialize dfu target to receive first image
+    return System::MapErrorZephyr(dfu_target_init(DFU_TARGET_IMAGE_TYPE_MCUBOOT, mCurrentImage, /* size */ 0, nullptr));
```
Comment (Damian-Nordic) [inline src/platform/nrfconnect/OTAImageProcessorImpl.cpp:50]: I think the first init should be called in `ProcessHeader` based on the image ID stored in the header. We should at least vierify that the first image is the app core image. `mCurrentImage` only store an image index in the OTA package, not the DFU target image ID.

File: src/platform/nrfconnect/OTAImageProcessorImpl.cpp [hunk 6]
```diff
@@ -76,23 +66,59 @@ CHIP_ERROR OTAImageProcessorImpl::Apply()
...
     VerifyOrReturnError(mDownloader != nullptr, CHIP_ERROR_INCORRECT_STATE);
 
     CHIP_ERROR error = ProcessHeader(block);
-
     if (error == CHIP_NO_ERROR)
     {
-        // DFU target library buffers data internally, so do not clone the block data.
-        error = System::MapErrorZephyr(dfu_target_write(block.data(), block.size()));
+        mParams.downloadedBytes += block.size();
+        if (mParams.downloadedBytes >= mContentHeader.mFiles[mCurrentImage].mFileSize)
```
Comment (Damian-Nordic) [inline src/platform/nrfconnect/OTAImageProcessorImpl.cpp:92]: I think you may want to accumulate the file sizes and calculate _file end offsets_ for each file. Otherwise, this condition will be needlessly triggered just after downloading the first block of the network image, right?

File: src/platform/nrfconnect/OTAImageProcessorImpl.cpp [hunk 7]
```diff
@@ -76,23 +66,59 @@ CHIP_ERROR OTAImageProcessorImpl::Apply()
...
+            // switch to next image
+            mCurrentImage++;
+            if (mContentHeader.mFiles[mCurrentImage].mFileSize > 0)
+            {
+                dfu_target_init(DFU_TARGET_IMAGE_TYPE_MCUBOOT, mCurrentImage, /* size */ 0, nullptr);
+                // write remaining data to new image
+                error =
+                    System::MapErrorZephyr(dfu_target_write(block.data() + (block.size() - remainedDataSize), remainedDataSize));
+                mParams.downloadedBytes = remainedDataSize;
+            }
```
Comment (Damian-Nordic) [inline src/platform/nrfconnect/OTAImageProcessorImpl.cpp:107]: `else` set error not to download any more blocks.

File: src/platform/nrfconnect/OTAImageProcessorImpl.cpp [hunk 8]
```diff
@@ -129,8 +155,15 @@ CHIP_ERROR OTAImageProcessorImpl::ProcessHeader(ByteSpan & block)
...
-        OTAImageContentHeader header = {};
-        CHIP_ERROR error             = mContentHeaderParser.AccumulateAndDecode(block, header);
+        CHIP_ERROR error = mContentHeaderParser.AccumulateAndDecode(block, mContentHeader);
+        ChipLogDetail(SoftwareUpdate, "Found following DFU Images:");
+        for (uint8_t i = 0; i < mContentHeader.kMaxFiles; i++)
+        {
+            ChipLogDetail(SoftwareUpdate, "[%d]: \n \
+                                           Image ID: %d \n \
+                                           Image size: %d",
+                          i, (uint32_t) mContentHeader.mFiles[i].mFileId, (uint32_t) mContentHeader.mFiles[i].mFileSize);
```
Comment (kkasperczyk-no) [inline src/platform/nrfconnect/OTAImageProcessorImpl.cpp:165]: Please use c++ style casting

**Review by kkasperczyk-no — APPROVED** [review]
I'm not really familiar with the OTA DFU core,  but the structure looks good to me overall. Please remember to update docker image number in the github workflow to use the new NCS version.

File: src/platform/nrfconnect/OTAImageProcessorImpl.cpp [hunk 9]
```diff
@@ -60,7 +61,15 @@ CHIP_ERROR OTAImageProcessorImpl::Abort()
 
 CHIP_ERROR OTAImageProcessorImpl::Apply()
 {
-    ReturnErrorOnFailure(System::MapErrorZephyr(dfu_target_done(true)));
+    int err = dfu_target_done(true);
+    if (err == 0)
+    {
+        err = dfu_target_schedule_update(-1);
```
Comment (markaj-nordic) [inline src/platform/nrfconnect/OTAImageProcessorImpl.cpp:67]: Would be good to know that -1 means.

Comment (LuDuda) [inline src/platform/nrfconnect/OTAImageProcessorImpl.cpp:67]: What if this line fails? Currently we're continue executing the code.

File: src/platform/nrfconnect/OTAImageProcessorImpl.cpp [hunk 10]
```diff
@@ -81,11 +90,41 @@ CHIP_ERROR OTAImageProcessorImpl::ProcessBlock(ByteSpan & block)
...
+                ChipLogDetail(SoftwareUpdate, "Remaining bytes from last block containing Net core data: %" PRIu64, remainingDataSize);
+                dfu_target_init(DFU_TARGET_IMAGE_TYPE_MCUBOOT, mCurrentImage.mIndex, /* size */ 0, nullptr);
+                // write remaining data to new image
+                error =
+                    System::MapErrorZephyr(dfu_target_write(block.data() + (block.size() - remainingDataSize), remainingDataSize));
+                mCurrentImage.mCurrentOffset = remainingDataSize;
+            }
+            else 
+            {
+                // set an error to end download process if the second image is not a Net core image.
```
Comment (LuDuda) [inline src/platform/nrfconnect/OTAImageProcessorImpl.cpp:117]: This comment may not be accurate, e.g. when other condition fails like `error` is not `CHIP_NO_ERROR` etc.

File: src/platform/nrfconnect/OTAImageProcessorImpl.cpp [hunk 11]
```diff
@@ -129,13 +168,21 @@ CHIP_ERROR OTAImageProcessorImpl::ProcessHeader(ByteSpan & block)
 
     if (mContentHeaderParser.IsInitialized() && !block.empty())
     {
-        OTAImageContentHeader header = {};
-        CHIP_ERROR error             = mContentHeaderParser.AccumulateAndDecode(block, header);
-
+        CHIP_ERROR error = mContentHeaderParser.AccumulateAndDecode(block, mContentHeader);
```
Comment (LuDuda) [inline src/platform/nrfconnect/OTAImageProcessorImpl.cpp:171]:
```suggestion
        CHIP_ERROR error = mContentHeaderParser.AccumulateAndDecode(block, mContentHeader);

```

File: src/platform/nrfconnect/OTAImageProcessorImpl.cpp [hunk 12]
```diff
@@ -81,11 +90,41 @@ CHIP_ERROR OTAImageProcessorImpl::ProcessBlock(ByteSpan & block)
...
+        if (mCurrentImage.mCurrentOffset >= mCurrentImage.mFileInfo->mFileSize)
+        {
+            // calculate how many data should be moved to the next image
+            uint64_t remainingDataSize = mCurrentImage.mCurrentOffset - static_cast<uint64_t>(mCurrentImage.mFileInfo->mFileSize);
+            // write last data of previous image
+            error = System::MapErrorZephyr(dfu_target_write(block.data(), block.size() - remainingDataSize));
+            // switch to net image
+            mCurrentImage.mIndex++;
+            mCurrentImage.mFileInfo = &mContentHeader.mFiles[mCurrentImage.mIndex];
+            if (OTAImageContentHeader::FileId::kNetMcuboot == mCurrentImage.mFileInfo->mFileId &&
```
Comment (LuDuda) [inline src/platform/nrfconnect/OTAImageProcessorImpl.cpp:105]:
```suggestion

            if (OTAImageContentHeader::FileId::kNetMcuboot == mCurrentImage.mFileInfo->mFileId &&
```

**Review by LuDuda — COMMENTED** [review]
Thanks :+1:

File: src/platform/nrfconnect/OTAImageProcessorImpl.cpp [hunk 13]
```diff
@@ -25,6 +25,7 @@
 #include <dfu/dfu_target.h>
 #include <dfu/dfu_target_mcuboot.h>
 #include <dfu/mcuboot.h>
+#include <logging/log.h>
```
Comment (Damian-Nordic) [inline src/platform/nrfconnect/OTAImageProcessorImpl.cpp:28]: Not needed?

File: src/platform/nrfconnect/OTAImageProcessorImpl.cpp [hunk 14]
```diff
@@ -81,11 +94,45 @@ CHIP_ERROR OTAImageProcessorImpl::ProcessBlock(ByteSpan & block)
...
     CHIP_ERROR error = ProcessHeader(block);
-
     if (error == CHIP_NO_ERROR)
     {
-        // DFU target library buffers data internally, so do not clone the block data.
-        error = System::MapErrorZephyr(dfu_target_write(block.data(), block.size()));
+        mCurrentImage.mCurrentOffset += block.size();
+        if (mCurrentImage.mCurrentOffset >= mCurrentImage.mFileInfo->mFileSize)
+        {
+            // calculate how many data should be moved to the next image
```
Comment (Damian-Nordic) [inline src/platform/nrfconnect/OTAImageProcessorImpl.cpp:102]:
```suggestion
            // calculate how much data should be moved to the next image
```

File: src/platform/nrfconnect/OTAImageProcessorImpl.cpp [hunk 15]
```diff
@@ -81,11 +94,45 @@ CHIP_ERROR OTAImageProcessorImpl::ProcessBlock(ByteSpan & block)
...
+            mCurrentImage.mFileInfo = &mContentHeader.mFiles[mCurrentImage.mIndex];
+
+            if (OTAImageContentHeader::FileId::kNetMcuboot == mCurrentImage.mFileInfo->mFileId &&
+                mCurrentImage.mFileInfo->mFileSize > 0 && CHIP_NO_ERROR == error)
+            {
+                // finish previous image and reset target
+                dfu_target_done(true);
+                dfu_target_reset();
+                // initialize next dfu target to store net-core image.
+                dfu_target_init(DFU_TARGET_IMAGE_TYPE_MCUBOOT, mCurrentImage.mIndex, /* size */ 0, nullptr);
```
Comment (Damian-Nordic) [inline src/platform/nrfconnect/OTAImageProcessorImpl.cpp:117]: We can do it as a followup, but imho we should check the error code here. Perhaps, extracting this part of the code to a separate function would help to organize the code that way (you could do early returns on error).
Also, using `mCurrentImage.mIndex` is misleading as those values are only accidentally the same. In theory you could have network image on a different position in the OTA image. You can simply use `1` or some constant.

**Review by Damian-Nordic — APPROVED** [review]

File: src/platform/nrfconnect/OTAImageProcessorImpl.cpp (hunk 15 above)
Comment (ArekBalysNordic) [reply src/platform/nrfconnect/OTAImageProcessorImpl.cpp:117]: You are right, this part can be moved to a separate function. mCurrentImage.mIndex contains information about current image which is not equal to mcuboot's image enumeration.  In mcuboot's image enum the mcuboot image is stored at 0, the app image is stored at 1 and the net core is stored at 2 so in this function it was needed to add another index. But I agree a constant will be better for our purpose because we don't support mcuboot image uptade yet.
